### ISSUE TITLE
Problem with agent with multiple objectives

### ISSUE THREAD

**Issue by jsalli** [Unity-Technologies/ml-agents#2533]
Hi,

I have modified the pushblock-example to have two types of blocks and goals. There are red blocks and blue blocks and also a red goal and a blue goal. Red block and goal are called in the code "bad block and bad goal" and the blue block and goal are called "good block and good goal".

The agent's objectives are to push the red blocks to red goal and the blue blocks to blue goal. 

The problem is that the agents seem to randonly choose to prefer one type of the two block types. So during different training sessions it might be that in one training session all agents push quickly the blue/good blocks to the blue/good goal but they perform badly with the red/bad block. On another training session the agents might choose to prefer the red/bad blocks and perform badly with the blue/good blocks.

I'm using curriculum learning with couple of options I've been playing around like randomly choosing only one type of block to the arena but nothing seems to work. [Here is the curriculum file](https://gitlab.com/jsalli/artificial-invaders-unity/blob/master/config/curricula/artificial-invader/ArtificialInvaderLearning.json)

My questions:
- Why are the agents not able to handle both types of blocks at the same time?
- Should I be using LSTM by setting the "use_recurrent"-option and it's sub options?
- Or should I train two brains so that one brain handles blue/good blocks and the other brain handles red/bad blocks and then switch between those brains in the code with some logic?

The project is available [here](https://gitlab.com/jsalli/artificial-invaders-unity) so you can try it yourself.

**Comment by awjuliani** [comment]
Hi @jsalli 

I would try thinking through the reward function a little. It may be the case that the agent prefers one or the other because there is not greater reward that would be achieved from equally preferring both. Perhaps you could try switching off between which goals are actually rewarding, so that the agent must learn to try both.

**Comment by jsalli** [comment]
Hi @awjuliani 
and thanks for the idea to change the rewarding. I made two branches in which I modified the rewarding system.
1) "dev/random-goal-type-gives-reward" in which the rewarding goal type is randomized on every agent reset and the rewarding goal gives +1 points and non-rewarding goal 0 points for correct goals. Scoring the block to the wrong goal resets the environment. I've also tested the game to end 1) if all scoring blocks are gone, 2) all blocks of both types are gone. In every setup the agents still choose randomly to prefer one type of block over the other.

2) "dev/last-goal-rewards-most". In this branch the agent is given a small rewards +0.1 points for all right goals and +1 point for the last right goal. Scoring the block to the wrong goal resets the environment. Still the agents choose to prefer randomly one type of block.

Are there any examples of agent having multiple objectives at the same time?

Is this issue most likely the rewarding system which leads the agent to find some unwanted way to maximize the reward?
